File: 1035-uncrossed-lines/1035-uncrossed-lines.cpp
class Solution {
public:
    
    int solve(int i,int j, vector<int>& nums1,vector<int>& nums2,vector<vector<int>>& dp){
        if(i<=0 || j<=0){
            return 0;
        }
        if(dp[i][j]!=-1){
            return dp[i][j];
        }
        
        
        if(nums1[i-1]==nums2[j-1]){
            dp[i][j] = 1 + solve(i-1,j-1,nums1,nums2,dp);
        } else {
            dp[i][j] = max(solve(i-1,j,nums1,nums2,dp),solve(i,j-1,nums1,nums2,dp));
        }
        
        return dp[i][j];
    }
    
    int maxUncrossedLines(vector<int>& nums1, vector<int>& nums2) {
        int n = nums1.size();
        int m = nums2.size();
        
        vector<vector<int>> dp(n+1,vector<int>(m+1,-1));
        
        return solve(n,m,nums1,nums2,dp);
    }
};

File: 1035-uncrossed-lines/1035-uncrossed-lines.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number}
 */
var maxUncrossedLines = function(nums1, nums2) {
    const m = nums1.length;
    const n = nums2.length;
    
    const dp = new Array(m+1);
    for (let i = 0; i <= m; i++) {
    dp[i] = new Array(n + 1).fill(0);
  }
    
    for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      if (nums1[i - 1] === nums2[j - 1]) {
        dp[i][j] = 1 + dp[i - 1][j - 1];
      } else {
        dp[i][j] = Math.max(dp[i][j - 1], dp[i - 1][j]);
      }
    }
  }
    return dp[m][n];
};